src/Microsoft.Framework.Runtime.Hosting/Loader/PackageAssemblyLoader.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Framework.Logging;
+using Microsoft.Framework.Runtime.Dependencies;
+using Microsoft.Framework.Runtime.Internal;
+using NuGet.Frameworks;
+using NuGet.ProjectModel;
+
+namespace Microsoft.Framework.Runtime.Loader
+{
+    /// <summary>
+    /// Loads .NET Assemblies from NuGet Packages
+    /// </summary>
+    /// <remarks>
+    /// This loader REQUIRES that a lock file has been generated for the project.
+    /// </remarks>
+    public class PackageAssemblyLoader : IAssemblyLoader
+    {
+        private readonly IAssemblyLoadContextAccessor _loadContextAccessor;
+        private readonly ILogger Log;
+        private readonly Dictionary<string, string> _assemblyLookupTable;
+
+        public PackageAssemblyLoader(IAssemblyLoadContextAccessor loadContextAccessor,
+            LockFile lockFile,
+            NuGetFramework targetFramework,
+            DefaultPackagePathResolver pathResolver)
+        {
+            _loadContextAccessor = loadContextAccessor;
+
+            Log = RuntimeLogging.Logger<PackageAssemblyLoader>();
+
+            // Initialize the assembly lookup table
+            _assemblyLookupTable = InitializeAssemblyLookup(lockFile, targetFramework, pathResolver);
+        }
+
+        /// <summary>
+        /// Loads an assembly from a NuGet package
+        /// </summary>
+        /// <remarks>
+        /// Note that the Package name is never used when loading assemblies.
+        /// If multiple packages provide the same assembly name, it is UNDEFINED
+        /// which assembly will be loaded.
+        /// </remarks>
+        /// <param name="name">The name of the assembly to load</param>
+        /// <returns>An <see cref="Assembly"/>, or null if the assembly could not be found</returns>
+        public Assembly Load(string name)
+        {
+            return Load(name, _loadContextAccessor.Default);
+        }
+
+        private Assembly Load(string name, IAssemblyLoadContext loadContext)
+        {
+            Log.LogVerbose($"Requested load of {name}");
+
+            string assemblyLocation;
+            if(_assemblyLookupTable.TryGetValue(name, out assemblyLocation))
+            {
+                return loadContext.LoadFile(assemblyLocation);
+            }
+            return null;
+        }
+
+        private Dictionary<string, string> InitializeAssemblyLookup(LockFile lockFile, NuGetFramework targetFramework, DefaultPackagePathResolver pathResolver)
+        {
+            var lookup = new Dictionary<string, string>();
+            var cacheResolvers = GetCacheResolvers();
+            foreach(var lib in lockFile.Libraries)
```

Comment:
You can even store the group resolved during the graph walk so you don't have to do it more than once per package.